### ISSUE TITLE
Support for configuring on a per-subdomain basis

### ISSUE THREAD

**Issue by timabbott** [django-auth-ldap/django-auth-ldap#27]
We've been using django-auth-ldap in [Zulip](https://zulipchat.com) happily for years.  As Zulip has grown, though, we've gotten requests for supporting LDAP authentication on our hosted service.

The most natural way to do this would be to have the django-auth-ldap settings configuration (things like `AUTH_LDAP_SERVER_URI`) be something that the LDAP backend looks up from e.g. database through some code that we write, rather than being something hardcoded in the Django `settings.py`.  Is there a good way to do that?

**Comment by jdufresne** [comment]
I think one approach would be to allow the [settings attribute](https://github.com/django-auth-ldap/django-auth-ldap/blob/c390d74a7cc6afe6601fa03f5bf3cbb8e33237cb/django_auth_ldap/backend.py#L115-L121) to be configurable. Right now, it is always `LDAPSettings`, if this could be overridden to a custom class, it could specify dynamic attributes. If you're interested, a PR implementing such an approach would be welcome. I'm also open to other ideas if you had something else in mind.

**Comment by jdufresne** [comment]
Just FYI, in the latest release, `AUTH_LDAP_SERVER_URI` can be a callable that accepts `request`. The callable could have logic that inspects the `request` object to decide where to connect. Does this solve your use case?

**Comment by timabbott** [comment]
@jdufresne thanks for the follow-up!  Unfortunately I think the answer is no, because we need all the other settings/configuration attributes to depend on the request as well (e.g. `AUTH_LDAP_USER_SEARCH`, `AUTH_LDAP_USER_ATTR_MAP`, `AUTH_LDAP_BIND_DN`, etc.).  Because if you're talking to multiple LDAP servers, they might be run by different people and/or have a different organizational structure.  

But if you're able to replace all of those settings with that model, I think that would probably solve my use case.

**Comment by timabbott** [comment]
Looking at the implementation for AUTH_LDAP_SERVER_URI, I think if we refactored this block:

```
             uri = self.settings.SERVER_URI
             if callable(uri):
-                uri = uri()
+                if func_supports_parameter(uri, 'request'):
+                    uri = uri(self._request)
+                else:
+                    warnings.warn(
+                        'Update AUTH_LDAP_SERVER_URI callable %s.%s to accept '
+                        'a positional `request` argument. Support for callables '
+                        'accepting no arguments will be removed in a future '
+                        'version.' % (uri.__module__, uri.__name__),
+                        DeprecationWarning
+                    )
+                    uri = uri()
```

to be a function like this:
```
def access_setting(self, setting_name):
   value = getattr(self.settings, setting_name)
   if callable(value):
       if func_supports_parameter(value, 'request'):
           value = value(self._request)
      ...
```

and then replaced access to `self.settings` with calls to that, it'd probably work.

(I'd want to refactor a bit more than that, at least to rename the current `self.settings` to `self._settings`, but that approach should be pretty doable)

**Comment by jdufresne** [comment]
I think that would work. Are you interested in creating a PR with tests to implement this idea?

**Comment by timabbott** [comment]
Yes, though it may take a while for me to find time personally; I'll post in our community about this idea and see if someone less busy is interested.

**Comment by timabbott** [comment]
@francoisfreitag since I think you're a new maintainer since the previous discussion, is this design something that you would accept a PR for? If so, do you have any guidance on how you'd like to see it done? It's starting to be problematic that Zulip can't support different LDAP configurations for different organizations, so we might find time to work on it soon. 

We've been otherwise quite happy with the library all these years.

**Comment by francoisfreitag** [comment]
Hi @timabbott,

I must be missing something, but wouldn’t it work for you to override `authenticate()` to do something like:
```python
class LDAPBackend:
    def authenticate(self, request, username=None, password=None, **kwargs):
        self.settings = get_settings(request)
        return super().authenticate(request, username=username, password=password, **kwargs)
```
